### ISSUE TITLE
Ostinato blocks on streams with MAC resolve

### ISSUE THREAD

**Issue by ehlers** [pstavirs/ostinato#186]
Ostinato: branch master, commit https://github.com/pstavirs/ostinato/commit/b559400cf19e4730dd9b23ebb7c05e87bb3fa465
OS: Tinycore Linux v6.4
Kernel 3.16.6 i686 (32bit)
gcc v4.9.1
protobuf v2.6.1
qt v4.8.6 from tinycore repository

I created a very simple test network, just ostinato connected to a router.
![network](https://cloud.githubusercontent.com/assets/896406/15738189/7250e330-28aa-11e6-8fdc-c64ab78817d1.png)

When I configure the device emulation everything looks fine. Ostinato resolves the router MAC address and the router shows the Ostinato mac address in it's arp cache. The eth1 interface in the ostinato PC has no interface configuration within linux, so the ARP resolution is really done by ostinato.
![device_information](https://cloud.githubusercontent.com/assets/896406/15738239/b7b25d96-28aa-11e6-8d6e-edeef927cd38.png)

Now I create a stream with protocols Ethernet II / IPv4 / ICMP and set the destination MAC address to "resolve".
![stream_mac](https://cloud.githubusercontent.com/assets/896406/15738331/0c5a414c-28ab-11e6-85a1-738a63aae5a4.png)
![stream_ipv4](https://cloud.githubusercontent.com/assets/896406/15738334/0ed88ffa-28ab-11e6-9872-acf22de038e4.png)

When I now press "Apply" to send this configuration to the drone, the GUI blocks (watch cursor). The drone process is still there (it doesn't terminate), it needs no CPU resources (not in an endless loop). The log file is attached.
[drone.log.txt](https://github.com/pstavirs/ostinato/files/295154/drone.log.txt)

When using a stream without the resolve feature, everything works fine.

**Comment by pstavirs** [comment]
It's embarrassing and bewildering how such a basic case does not work!

I was not able to reproduce the problem on my main development environment where I did all the development for this feature - Windows.

But I saw it instantly on Linux! After much debugging, it turns out it isn't really the platform difference but a difference in a Qt API behaviour across Qt versions. Try the following patch -
```diff
diff --git a/server/myservice.cpp b/server/myservice.cpp
index b707896..d6f79da 100644
--- a/server/myservice.cpp
+++ b/server/myservice.cpp
@@ -50,7 +50,7 @@ MyService::MyService()

     for (int i = 0; i < n; i++) {
         portInfo.append(portManager->port(i));
-        portLock.append(new QReadWriteLock());
+        portLock.append(new QReadWriteLock(QReadWriteLock:Recursive));
     }
 }
```
This may or may not be the final fix, but it should unblock you from further testing this feature.

**Comment by ehlers** [comment]
Your fix doesn't compile with my gcc, changed `QReadWriteLock:Recursive` to `QReadWriteLock::Recursive` (note the double colon). Now it compiles (and more important) my test works now. Will try some more advanced tests tomorrow.

**Comment by ehlers** [comment]
Tested device emulation with IPv4 and IPv6, VLANs, multiple devices per VLAN and multiple device groups on an interface. No issues found, that's really a great stuff.

Perhaps just one small idea for improvement:
After setting up a device group the functions to start ARP/ND resolution are in the lower part of the GUI and the window to display the result are in the upper part. What about adding the "Resolve Neighbors" and "Clear Neighbors" buttons also in the device information window? That way initiating the neighbor resolution and having the information about the results are next to each other.

**Comment by pstavirs** [comment]
Can you raise a new issue for the last suggestion and I will investigate the feasibility of doing that?
